Add basic support for check constraints to database migrations

fatkodima opened the pull request:
This patch adds basic support for adding and removing check constraints using the migration DSL:
```
create_table :distributors do |t|
  t.string :zipcode
  t.check_constraint "zipchk", "char_length(zipcode) = 5"   
end

remove_check_constraint :distributors, "zipchk"
```

Check constraints are dumped inside `create_table` sections to `schema.rb`.

`add_check_constraint` and `remove_check_constraint` are both reversible.

### Considerations
- Maybe remove the `_constraint` part from new methods names? Seems a little redundant
- Add additional options, such as `NO INHERIT`(PostgreSQL)?

Related https://github.com/rails/rails/issues/23083.

fatkodima committed:
Add basic support for check constraints to database migrations

sgrif commented on activerecord/lib/active_record/migration/command_recorder.rb:
```diff
@@ -112,14 +113,15 @@ def change_table(table_name, options = {}) # :nodoc:
-              add_timestamps:    :remove_timestamps,
-              add_reference:     :remove_reference,
-              enable_extension:  :disable_extension
+            { transaction:           :transaction,
```
Why was every line in this hash changed when you only added a single line?

matthewd replied:
@sgrif  because it's very clearly [already] aligned to the longest key?

(though it does seem to have ended up one space further over than it should be.)

sgrif replied:
Sorry, I should have been more clear on my intent:

Do you think we could drop the visual indent change here so we aren't changing every line in this hash? I think having the one line that is not indented the same is fine.

sgrif reviewed (changes requested):
I don't think the tests here are sufficient. There are no tests which check that the constraint has actually been added to the table (which can only really be tested by attempting an insertion that should fail).

In particular, I don't think you've handled SQLite correctly, where check constraints are supported, but can only be added inline with the column definition. Unless I'm missing something, there's no way to add a check constraint as part of a column definition in this PR at all.

lostapathy commented:
+1 to @sgrif's thoughts on sqlite.  I also recently discovered that add_reference doesn't create foreign keys on sqlite, even though sqlite could support it.  I'm working on a PR for that, and it'd great to not see sqlite fall behind farther.

matthewd commented:
> Unless I'm missing something, there's no way to add a check constraint as part of a column definition in this PR at all.

That sounds more like an additional feature than a flaw in this PR, per se.

(i.e., if the PR were implementing column-level constraints and had omitted SQLite, that would be more of a problem; whereas I think only implementing table-level constraints is a perfectly reasonable self-contained feature, particularly as someone's first substantial contribution.)

sgrif commented:
I'm fine with omitting SQLite support entirely for this feature to start, but I still think we need tests which actually exercise the added constraint.

lostapathy commented:
@matthewd you're right, that makes sense. This API makes sense in general, just not for sqlite. Totally fair to skip it.  Much like how add_foreign_key can't work on sqlite, but add_reference should be doable.

Would it be possible/tolerable to emit some kind of warning if a constraint (or foreign key) migration is treated as a no-op sqlite3?  I spent a ton of time tracing that behavior out and I'm not sure what we do now is intuitive.  The sqlite docs say they support foreign keys and foreign key migrations are "successful", so it one does not expect to end up without a foreign key in their sqlite db.

fatkodima commented:
Updated PR. Added missing test (suggested by @sgrif ) and support for sqlite.

pkoch commented:
What's missing to move forward on this branch atm?

NikolayS commented:
This is a good feature. It's 2019 already, while `CHECK` is from SQL-92. Good time to start supporting it, both for MySQL and Rails.

fatkodima commented:
If core team will accept this PR, I will update it as needed.
Btw, right now I think it is a bad idea to support constraints, because currently migrations are not far from being bloated and there will be people who wants to support other not so popular features like triggers and other db stuff.

NikolayS commented:
Constraints enforcement is a database system's job.

In general, only DBMS is able enforce constraints in 100% cases. If you implement checks in your app logic, there are significant chances that after one year or so they will be silently broken, because some human accessed DB directly, changing data, or because you've start using API in different language/framework, or because you've used some DB migrations bypassing your abstraction layer which tried to enforce constraints.

`CHECK` is a very good thing, underestimated due to lack of its support in MySQL (until 8.0.16) and lack of its support in frameworks. This is a class problem "chicken vs egg" problem: without support in frameworks, people won't start using it.

fatkodima commented:
@matthewd Wdyt about moving this forward?

pkoch commented:
@matthewd @sgrif I'd love to not let this die. What would it take to get a ship it?

jeremy commented on activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb:
```diff
@@ -662,21 +685,39 @@ def remove_references(*args, **options)
+      #  t.foreign_key(:authors) unless t.foreign_key_exists?(:authors)
       #
       # See {connection.foreign_key_exists?}[rdoc-ref:SchemaStatements#foreign_key_exists?]
-      def foreign_key_exists?(*args) # :nodoc:
```
Related to this PR?

jeremy commented on activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb:
```diff
@@ -1034,6 +1034,49 @@ def foreign_key_options(from_table, to_table, options) # :nodoc:
+      # Returns an array of check constraints for the given table.
+      # The check constraints are represented as CheckConstraintDefinition objects.
+      def check_constraints(table_name)
+        raise NotImplementedError, "check_constraints is not implemented"
```
Can omit the message; redundant.

jeremy commented on activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb:
```diff
@@ -1034,6 +1034,49 @@ def foreign_key_options(from_table, to_table, options) # :nodoc:
+        if check_constraints.detect { |i| i.name == constraint_name.to_s }
+          _remove_check_constraint(table_name, constraint_name)
+        else
+          raise ArgumentError, "Check constraint name '#{constraint_name}' on table '#{table_name}' not exists"
```
`not exists` → `does not exist`

jeremy commented on activerecord/test/cases/migration/check_constraint_test.rb:
```diff
@@ -0,0 +1,104 @@
+          assert_equal "trades", constraint.table_name
+          assert_equal "quantity_check", constraint.name
+          assert_equal expression, constraint.expression
+        end
```
Test coverage for the `change_table { |t| t.check_constraint :name, "expression" }` form as well?

jeremy reviewed (commented):
🙌🏼

**Method naming**
> Maybe remove the _constraint part from new methods names? Seems a little redundant

Comparing,

```ruby
create_table :distributors do |t|
  t.string :zipcode
  t.check_constraint "zipchk", "char_length(zipcode) = 5"   
end

remove_check_constraint :distributors, "zipchk"
```

```ruby
create_table :distributors do |t|
  t.string :zipcode
  t.check "zipchk", "char_length(zipcode) = 5"   
end

remove_check :distributors, "zipchk"
```

I prefer the clarity of seeing constraint mentioned. If it were a constraint on a column definition, `check: "expression"` would be nicer.

**Naming constraints**
Having to name constraints is a hassle, too. Can we name them automatically, like we do with foreign keys and indexes? For example, we could transform the check expression into name like `chk_rails_#{hashed_identifier}`.

fatkodima commented:
1. Updated this with current master branch.
2. fixed @jeremy suggestions (including optional constraint name feature)
3. added missing implementation for mysql

jeremy reviewed (approved):
Ready to merge pending rebase. Thanks for your work on this @fatkodima.

fatkodima commented:
@jeremy Rebased.

NikolayS commented:
Wow. Congrats :)
